Source/WebCore/rendering/RenderLayerScrollableArea.cpp

```diff
@@ -845,13 +845,19 @@ bool RenderLayerScrollableArea::canShowNonOverlayScrollbars() const
     return canHaveScrollbars() && !(m_layer.renderBox() && m_layer.renderBox()->canUseOverlayScrollbars());
 }
 
+void RenderLayerScrollableArea::updateScrollbarsControllerAfterLayerCreation()
+{
+    if (auto scrollbarController = m_layer.page().chrome().client().createScrollbarsController(m_layer.page(), *this)) {
```

Comment:
You've still got some "implicit" handshakes in this code; `WebChromeClient::createScrollbarsController()` checks whether the ScrollableArea does async scrolling, and returns null. So someone implementing `createScrollbarsController()` in their own chrome client would have to know that there's a "return null means no change" contract here.

It's also weird for ScrollableArea to expose an implementation detail (`hasRemoteScrollbarsController()`).

I think a better approach would be to have ChromeClient implement:
`updateScrollbarsControllerIfNecessary()`, and have it get the `ScrollbarsController` from the `ScrollableArea`, and then ask about its type.

Reply:
Fixed.